data/base/endpoint.php

```diff
@@ -0,0 +1,280 @@
+	 *
+	 * @return string current location will later be added automatically.
+	 */
+	public static function get_format() {
+		return '';
+	}
+
+	/**
+	 * Get base route.
+	 *
+	 * Removing 'index' from endpoint.
+	 *
+	 * @return string
+	 */
+	protected function get_base_route() {
+		$endpoint_name = $this->get_name();
+
+		// TODO: Allow this only for internal routes.
+		// TODO: Make difference between internal and external endpoints.
+		if ( 'index' === $endpoint_name ) {
+			$endpoint_name = '';
+		}
+
+		return '/' . $this->controller->get_rest_base() . '/' . $endpoint_name;
+	}
+
+	/**
+	 * Register the endpoint.
+	 *
+	 * By default: register get items route.
+	 *
+	 * @throws \Exception
+	 */
+	protected function register() {
+		$this->register_items_route();
+	}
+
+	/**
+	 * Register sub endpoint.
+	 *
+	 * @param string $route
+	 * @param string $endpoint_class
+	 *
+	 * @throws \Exception
+	 */
+	protected function register_sub_endpoint( $route, $endpoint_class ) {
+		$endpoint_instance = new $endpoint_class( $route, $this );
+
+		if ( ! ( $endpoint_instance instanceof SubEndpoint ) ) {
+			throw new \Exception( 'Invalid endpoint instance.' );
+		}
+
+		$endpoint_route = $route . '/' . $endpoint_instance->get_name();
+
+		$this->sub_endpoints[ $endpoint_route ] = $endpoint_instance;
+
+		$component_name = $endpoint_instance->controller->get_rest_base();
+		$parent_instance = $endpoint_instance->get_parent();
+		$parent_name = $endpoint_instance->get_name();
+		$parent_format_suffix = $parent_instance::get_format();
+		$current_format_suffix = $endpoint_instance::get_format();
+
+		$command = $component_name . '/' . $parent_name;
+		$format = $component_name . '/' . $parent_format_suffix . '/' . $parent_name . '/' . $current_format_suffix;
+
+		$this->controller->register_endpoint_format( $command, $format );
+	}
+
+	/**
+	 * Retrieves a collection of items.
+	 *
+	 * @param \WP_REST_Request $request Full data about the request.
+	 *
+	 * @return \WP_Error|\WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	protected function get_items( $request ) {
+		return $this->controller->get_items( $request );
+	}
+
+	/**
+	 * Retrieves one item from the collection.
+	 *
+	 * @param string $id
+	 * @param \WP_REST_Request $request Full data about the request.
+	 *
+	 * @return \WP_Error|\WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	protected function get_item( $id, $request ) {
+		return $this->controller->get_item( $request );
+	}
+
+	/**
+	 * Retrieves a recursive collection of all endpoint(s), items.
+	 *
+	 * Get items recursive, will run overall endpoints of the current controller.
+	 * For each endpoint it will run `$endpoint->getItems( $request ) // the $request passed in get_items_recursive`.
+	 * Will skip self endpoint ( more information available a 'test-endpoint.php'.
```

Comment:
Please don't refer to the tests. add an example here.

Reply:
Handled: https://github.com/elementor/elementor/pull/11388/commits/d2975cde01e3e6b6ba526c44fb501c407482c2ec